Remove use of SIMPL_* macros and replace with modern C++ code

imikejackson opened the pull request:
A number of things were done with this branch.

* All uses of SIMPL_* macros were replaced with real code
* #ifdef around the pybind11 sections
* Pybind11 macros move to their own header
* #include DataContainerArray.h was replace with a forward declaration in AbstractFilter
* Filters and other classes updated with needed include directives.

imikejackson committed:
Add compiler define if SIMPL_ENABLE_PYTHON is turned on.

Update clang-format to NOT sort the includes.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Wrap the Python wrapping macros in a #define for SIMPL_ENABLE_PYTHON

This is done because SIMPLibSetGetMacros is going away eventually and we need to separate
out the still needed macros that help generate the python bindings.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Massive code modifications to clean up and reduce macro clutter

* Remove use of all SIMPL_* macros replacing with actual C++ code
* Remove the include directive to DataContainerArray in AbstractFilter instead using a forward declaration and alias
* Add needed include directives or forward declarations to all effected classes

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Transfer some SIMPL_* macros into the Pybind11 macro section

This is needed because the codescraper algorithm looks for four specific macros in order
to determine the proper c++ and python code to generate. The same macros have been
defined in the SIMPLPythonMacros.h header file.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented:
Assuming the PyBind macros don't force the issue, it might be good to search and replace `const T&` with `T` for the following types:

- int
- float
- double
- bool
- char
- ArrayCalculator::AngleUnit
- SIMPL::ScalarTypes::Type
- SIMPL::NumericTypes::Type

imikejackson committed:
Fix pass by const reference for POD types

In the last pass at removing the macros POD types were mistakenly passed by
const reference instead of by value. These have all been fixed to pass by value.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/SIMPLib/CoreFilters/Testing/Cxx/ArrayCalculatorTest.cpp:
```diff
@@ -53,13 +53,32 @@
+  /**
+   * @brief Returns the name of the class for DummyObserver
+   */
+  /**
```
Something odd happened here, and you've got duplicate comment blocks for this method.

mmarineBlueQuartz commented on Source/SIMPLib/CoreFilters/Testing/Cxx/MultiThresholdObjects2Test.cpp:
```diff
@@ -49,13 +48,32 @@
-  SIMPL_TYPE_MACRO(MultiThresholdObjects2Test)
+  /**
+   * @brief Returns the name of the class for MultiThresholdObjects2Test
+   */
```
Duplicate method descriptions again.

mmarineBlueQuartz commented on Source/SIMPLib/CoreFilters/Testing/Cxx/MultiThresholdObjectsTest.cpp:
```diff
@@ -47,13 +46,32 @@
-  SIMPL_TYPE_MACRO(MultiThresholdObjectsTest)
+  /**
+   * @brief Returns the name of the class for MultiThresholdObjectsTest
+   */
```
And another duplicate description.

mmarineBlueQuartz replied:
This seems to be the case for much of the Testing/Cxx directory.  Consider this a recurring comment rather than making note of all the rest of the test classes.

mmarineBlueQuartz commented on Source/SIMPLib/CoreFilters/util/ABSOperator.cpp:
```diff
@@ -58,7 +58,11 @@ ABSOperator::~ABSOperator() = default;
 // -----------------------------------------------------------------------------
-void ABSOperator::calculate(AbstractFilter* filter, DataArrayPath calculatedArrayPath, QStack<ICalculatorArray::Pointer>& executionStack)
+void ABSOperator::calculate(AbstractFilter* filter, DataArrayPath calculatedArrayPath,
+                            QStack<ICalculatorArray::Pointer>& executionStack){CREATE_NEW_ARRAY_STANDARD_UNARY(filter, calculatedArrayPath, executionStack, fabs)}
```
I don't think clang-format did this, but if it did, something needs to be tweaked.  I more than half-expected this line to be a lambda and was wondering where the `[=]` and closing parenthesis were.

mmarineBlueQuartz replied:
This seems to be a recurring problem in ArrayCalculator's operators for this same method.

mmarineBlueQuartz commented on Source/SIMPLib/CoreFilters/util/ICalculatorArray.h:
```diff
@@ -38,6 +38,10 @@
 
 #include "SIMPLib/DataArrays/DataArray.hpp"
 #include "SIMPLib/SIMPLib.h"
+#include "SIMPLib/DataArrays/DataArray.hpp"
```
Duplicate DataArray includes

mmarineBlueQuartz commented on Source/SIMPLib/TestFilters/FilterGroup00.cpp:
```diff
@@ -187,3 +189,32 @@ const QString TESTCLASSNAME::getHumanLabel() const
+// -----------------------------------------------------------------------------
+std::shared_ptr<TESTCLASSNAME> FilterGroup00::New()
+{
+  struct make_shared_enabler : public TESTCLASSNAME
```
Use of TESTCLASSNAME instead of FilterGroup00 implies that FilterGroup00 may have been mass replaced by accident.  I strongly suggest looking at the .h / .cpp pair and confirm that everything is as expected.

mmarineBlueQuartz commented on Source/SVWidgetsLib/FilterParameterWidgets/DataContainerArrayProxyWidget.cpp:
```diff
@@ -37,10 +37,13 @@
+#include <QtCore/QDebug>
+
+#include "SIMPLib/DataContainers/DataContainerArray.h"
+#include "SIMPLib/DataContainers/DataContainerArray.h"
```
Duplicate DataContainerArray includes

mmarineBlueQuartz reviewed (changes requested):
- FilterGroup00 needs to be examined to make sure a text substitution didn't go off the rails and rename the class unexpectedly.
- ArrayCalculator operator classes require updated formatting for their calculate methods.  Splitting the parameters between multiple lines but reducing the method brackets to one line alongside half the parameters makes it less clear than it should be if that is the method definition or an oddly placed lambda.

Odd but not important
- SIMPL_TYPE_MACRO(T) swapped out with duplicate method descriptions.
- Duplicated include paths in the same file.

imikejackson committed:
Fixing code based on feed back from PR

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Resolving merge conflicts

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix more pass by reference for POD types

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/Experimental/FilterCallback/FilterParameter.h:
```diff
@@ -258,6 +334,14 @@ class DataArrayPathParameter : public IFilterParameter
+    int m_GroupIndex = {};
+    int m_DefaultValue = {};
+    double m_DefaultValue = {};
+    DataArrayPath m_DefaultValue = {};
```
I know this is in experimental, but I don't think you can have multiple variables of the same name and different types.

mmarineBlueQuartz commented on Source/SIMPLib/Common/CreatedArrayHelpIndexEntry.cpp:
```diff
@@ -47,3 +47,136 @@ CreatedArrayHelpIndexEntry::CreatedArrayHelpIndexEntry()
+}
+
+// -----------------------------------------------------------------------------
+const QString CreatedArrayHelpIndexEntry::getNameOfClass() const
```
Unnecessary const in the beginning

imikejackson replied:
Those should be taken care of

imikejackson committed:
Merge and Compile error fixes

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix warnings about returning const QString as a value instead of a reference.

Warnings are fixed by changing from "const QString" to "QString"

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Update Messages classes and remove Python Wrapping from pure virtual class

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Add a clang pragma to quiet a warning about using std::move.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Improve generated CPP code to modern standards

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix compiler warnings about overriding a virtual method without using override keyword

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix DOxygen comments that got munged up in some find/replace operation

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fixing compiler errors on MSVC Platform

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Add conditional #pragma for clang diagnostics

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/SIMPLib/Common/Constants.h:
```diff
@@ -53,30 +65,29 @@
+const QString PathSep("|");
+static const uint8_t Unchecked = 0;
+static const uint8_t PartiallyChecked = 1;
+static const uint8_t Checked = 2;
```
`static const` global

imikejackson replied:
As we discussed, we are going to leave these for the time being until we can get on C++17 and use inline keyword

mmarineBlueQuartz commented on Source/SIMPLib/Common/Constants.h:
```diff
@@ -53,30 +65,29 @@
+{
+static const unsigned int LowtoHigh = 0;
+static const unsigned int HightoLow = 1;
+static const unsigned int UnknownRefFrameZDirection = 2;
```
`static const` globals

imikejackson committed:
Address feedback from code review.

Remove static const from constants.h
Fix duplicate variable names in experimental file.

imikejackson committed:
Feature/const (#380)

* Added const to appropriate DataArray functions

* NeighborList::writeH5Data needed some additional changes since it was
trying to modify a member variable

* Added const to appropriate Geometry functions

* LinkedDataContainerSelectionWidget needed the IGeometry include to work

* Added const to DynamicTableData::getTableData()

* Added const to appropriate functions in AM, DC, and DCA classes

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix compiler errors due to PYBIND_11 macros not being updated to API changes

* develop:
  Fix compiler errors due to PYBIND_11 macros not being updated to API changes
  Feature/const (#380)

# Conflicts:
#	Source/SIMPLib/DataArrays/StatsDataArray.h
#	Source/SIMPLib/DataArrays/StringDataArray.h
#	Source/SIMPLib/DataContainers/AttributeMatrix.h
#	Source/SIMPLib/DataContainers/DataContainerArray.h
#	Source/SIMPLib/Geometry/IGeometry.h
#	Source/SIMPLib/Geometry/IGeometry2D.h
#	Source/SIMPLib/Geometry/IGeometryGrid.h
#	Source/SIMPLib/Geometry/ImageGeom.h
#	Source/SVWidgetsLib/FilterParameterWidgets/LinkedDataContainerSelectionWidget.cpp

imikejackson committed:
Update codes to latest const correctness changes

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Add missing include directive <memory> for shared_ptr<> access.

Updated some other code areas with better documentation strings
Updated some code sections to be more consistent with use of type aliases.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix bug using std::accumulate due to casting issue.

Passing in "1" will force the computation to be done in int32_t. When using large
Tuple dimensions we will over flow.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>
